File: vidiaspot-marketplace/frontend/src/services/deliveryService.js
// Delivery Management Service for Web Application

// Mock API service for delivery driver functionality
class DeliveryService {
  constructor() {
    // In a real app, this would connect to the backend API
    this.baseURL = '/api/delivery'; // Placeholder for actual API
  }

  // Fetch delivery packages for driver
  async getDeliveryPackages(driverId) {
    // Simulate API call with mock data
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve([
          {
            id: 'pkg-1',
            trackingNumber: 'VID001234567',
            recipientName: 'John Doe',
            recipientPhone: '+2348012345678',
            deliveryAddress: '123 Main Street, Lagos',
            packageDetails: 'Electronics - Phone',
            weight: 2.5,
            volume: 0.01,
            requiresSignature: true,
            requiresPhoto: true,
            requiresIdCheck: false,
            priority: 2,
            scheduledDeliveryTime: new Date(),
            deliveredAt: null,
            eta: new Date(Date.now() + 30 * 60000), // 30 minutes from now
            status: 'outForDelivery', // pending, inTransit, outForDelivery, delivered, failed, returned
            signature: null,
            photo: null,
            notes: ''
          },
          {
            id: 'pkg-2',
            trackingNumber: 'VID001234568',
            recipientName: 'Jane Smith',
            recipientPhone: '+2348012345679',
            deliveryAddress: '456 Business Avenue, Lagos',
            packageDetails: 'Documents',
            weight: 0.5,
            volume: 0.002,
            requiresSignature: true,
            requiresPhoto: false,
            requiresIdCheck: true,
            priority: 1,
            scheduledDeliveryTime: new Date(),
            deliveredAt: null,
            eta: new Date(Date.now() + 45 * 60000), // 45 minutes from now
            status: 'pending',
            signature: null,
            photo: null,
            notes: ''
          },
          {
            id: 'pkg-3',
            trackingNumber: 'VID001234569',
            recipientName: 'Bob Johnson',
            recipientPhone: '+2348012345670',
            deliveryAddress: '789 Innovation Hub, Lagos',
            packageDetails: 'Clothing items',
            weight: 1.2,
            volume: 0.008,
            requiresSignature: false,
            requiresPhoto: true,
            requiresIdCheck: false,
            priority: 3,
            scheduledDeliveryTime: new Date(),
            deliveredAt: null,
            eta: new Date(Date.now() + 60 * 60000), // 60 minutes from now
            status: 'inTransit',
            signature: null,
            photo: null,
            notes: ''
          }
        ]);
      }, 500); // Simulate network delay
    });
  }

  // Update package status
  async updatePackageStatus(packageId, status, data = {}) {
    // Simulate API call
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve({
          success: true,
          message: 'Package status updated successfully',
          packageId,
          status,
          data
        });
      }, 300);
    });
  }

  // Mark package as delivered
  async markPackageAsDelivered(packageId, deliveryData) {
    // Simulate API call
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve({
          success: true,
          message: 'Package marked as delivered',
          packageId,
          deliveredAt: new Date()
        });
      }, 500);
    });
  }

  // Get optimized route
  async getOptimizedRoute(driverId) {
    // Simulate API call
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve([
          {
            id: 'stop-1',
            address: '123 Main Street, Lagos',
            customerName: 'John Doe',
            packageDetails: 'Electronics - Phone',
            priority: 2,
            weight: 2.5,
            volume: 0.01,
            timeWindowStart: '09:00',
            timeWindowEnd: '11:00',
            isCompleted: false,
            latitude: 6.456,
            longitude: 3.387
          },
          {
            id: 'stop-2', 
            address: '456 Business Avenue, Lagos',
            customerName: 'Jane Smith',
            packageDetails: 'Documents',
            priority: 1,
            weight: 0.5,
            volume: 0.002,
            timeWindowStart: '10:00',
            timeWindowEnd: '12:00',
            isCompleted: false,
            latitude: 6.432,
            longitude: 3.398
          },
          {
            id: 'stop-3',
            address: '789 Innovation Hub, Lagos',
            customerName: 'Bob Johnson',
            packageDetails: 'Clothing items',
            priority: 3,
            weight: 1.2,
            volume: 0.008,
            timeWindowStart: '11:00',
            timeWindowEnd: '13:00',
            isCompleted: true,
            latitude: 6.445,
            longitude: 3.376
          }
        ]);
      }, 500);
    });
  }

  // Get driver conversations
  async getConversations(driverId) {
    // Simulate API call
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve([
          {
            id: 'conv-1',
            customerId: 'cust-1',
            customerName: 'John Doe',
            lastMessage: 'Hi, I will be home between 2pm and 4pm today',
            lastMessageAt: new Date(Date.now() - 30 * 60000), // 30 minutes ago
            unreadCount: 1,
            messages: [
              {
                id: 'msg-1',
                sender: 'driver',
                content: 'Hi John, I will be delivering your package between 2pm and 4pm today',
                timestamp: new Date(Date.now() - 60 * 60000) // 1 hour ago
              },
              {
                id: 'msg-2',
                sender: 'customer',
                content: 'Hi, I will be home between 2pm and 4pm today',
                timestamp: new Date(Date.now() - 30 * 60000) // 30 minutes ago
              }
            ]
          },
          {
            id: 'conv-2',
            customerId: 'cust-2',
            customerName: 'Jane Smith',
            lastMessage: 'Can you please reschedule for tomorrow?',
            lastMessageAt: new Date(Date.now() - 2 * 60 * 60000), // 2 hours ago
            unreadCount: 0,
            messages: [
              {
                id: 'msg-3',
                sender: 'driver',
                content: 'Your delivery is scheduled for today between 10am and 12pm',
                timestamp: new Date(Date.now() - 3 * 60 * 60000) // 3 hours ago
              },
              {
                id: 'msg-4',
                sender: 'customer',
                content: 'Can you please reschedule for tomorrow?',
                timestamp: new Date(Date.now() - 2 * 60 * 60000) // 2 hours ago
              }
            ]
          }
        ]);
      }, 500);
    });
  }

  // Send message to customer
  async sendMessage(conversationId, message) {
    // Simulate API call
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve({
          success: true,
          message: 'Message sent successfully',
          conversationId,
          sentMessage: message
        });
      }, 300);
    });
  }

  // Update driver status
  async updateDriverStatus(driverId, status) {
    // Simulate API call
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve({
          success: true,
          message: 'Driver status updated successfully',
          driverId,
          status
        });
      }, 300);
    });
  }
}

export default new DeliveryService();

File: vidiaspot-marketplace/mobile_delivery_driver/lib/services/delivery_management/delivery_service.dart
import 'package:flutter/foundation.dart';
import 'package:geolocator/geolocator.dart';

class DeliveryManagementService extends ChangeNotifier {
  List<DeliveryPackage> _packages = [];
  List<DeliveryEvent> _deliveryEvents = [];
  
  List<DeliveryPackage> get packages => _packages;
  List<DeliveryEvent> get deliveryEvents => _deliveryEvents;
  
  // Add a package to the delivery list
  void addPackage(DeliveryPackage deliveryPackage) {
    _packages.add(deliveryPackage);
    notifyListeners();
  }
  
  // Update package status
  void updatePackageStatus(String packageId, DeliveryStatus status) {
    int index = _packages.indexWhere((pkg) => pkg.id == packageId);
    if (index != -1) {
      _packages[index] = _packages[index].copyWith(status: status);
      
      // Add delivery event
      _deliveryEvents.add(DeliveryEvent(
        id: DateTime.now().millisecondsSinceEpoch.toString(),
        packageId: packageId,
        status: status,
        timestamp: DateTime.now(),
        location: geolocatorPosition,
      ));
      
      notifyListeners();
    }
  }
  
  // Mark package as delivered with proof
  void markPackageAsDelivered(String packageId, {String? signature, String? photo, String? notes}) {
    int index = _packages.indexWhere((pkg) => pkg.id == packageId);
    if (index != -1) {
      _packages[index] = _packages[index].copyWith(
        status: DeliveryStatus.delivered,
        deliveredAt: DateTime.now(),
        signature: signature,
        photo: photo,
        notes: notes,
      );
      
      // Add delivery event
      _deliveryEvents.add(DeliveryEvent(
        id: DateTime.now().millisecondsSinceEpoch.toString(),
        packageId: packageId,
        status: DeliveryStatus.delivered,
        timestamp: DateTime.now(),
        location: geolocatorPosition,
        signature: signature,
        photo: photo,
        notes: notes,
      ));
      
      notifyListeners();
    }
  }
  
  // Get packages by status
  List<DeliveryPackage> getPackagesByStatus(DeliveryStatus status) {
    return _packages.where((pkg) => pkg.status == status).toList();
  }
  
  // Get packages by priority
  List<DeliveryPackage> getPackagesByPriority(int priority) {
    return _packages.where((pkg) => pkg.priority == priority).toList();
  }
  
  // Get ETA for a package
  DateTime? getEtaForPackage(String packageId) {
    int index = _packages.indexWhere((pkg) => pkg.id == packageId);
    if (index != -1) {
      return _packages[index].eta;
    }
    return null;
  }
  
  // Update ETA for a package
  void updateEtaForPackage(String packageId, DateTime newEta) {
    int index = _packages.indexWhere((pkg) => pkg.id == packageId);
    if (index != -1) {
      _packages[index] = _packages[index].copyWith(eta: newEta);
      notifyListeners();
    }
  }
}

// Placeholder for geolocator position
Position geolocatorPosition = Position(
  latitude: 0.0,
  longitude: 0.0,
  timestamp: DateTime.now(),
  accuracy: 0.0,
  altitude: 0.0,
  heading: 0.0,
  speed: 0.0,
  speedAccuracy: 0.0,
);

enum DeliveryStatus {
  pending,
  inTransit,
  outForDelivery,
  delivered,
  failed,
  returned,
}

class DeliveryPackage {
  final String id;
  final String trackingNumber;
  final String recipientName;
  final String recipientPhone;
  final String deliveryAddress;
  final double latitude;
  final double longitude;
  final String packageDetails;
  final double weight; // in kg
  final double volume; // in cubic meters
  final bool requiresSignature;
  final bool requiresPhoto;
  final bool requiresIdCheck;
  final int priority; // 1 (highest) to 5 (lowest)
  final DateTime? scheduledDeliveryTime;
  final DateTime? deliveredAt;
  final DateTime? eta;
  final DeliveryStatus status;
  final String? signature;
  final String? photo;
  final String? notes; // delivery notes or exceptions
  
  DeliveryPackage({
    required this.id,
    required this.trackingNumber,
    required this.recipientName,
    required this.recipientPhone,
    required this.deliveryAddress,
    required this.latitude,
    required this.longitude,
    required this.packageDetails,
    this.weight = 0.0,
    this.volume = 0.0,
    this.requiresSignature = true,
    this.requiresPhoto = true,
    this.requiresIdCheck = false,
    this.priority = 3,
    this.scheduledDeliveryTime,
    this.deliveredAt,
    this.eta,
    this.status = DeliveryStatus.pending,
    this.signature,
    this.photo,
    this.notes,
  });
  
  DeliveryPackage copyWith({
    String? id,
    String? trackingNumber,
    String? recipientName,
    String? recipientPhone,
    String? deliveryAddress,
    double? latitude,
    double? longitude,
    String? packageDetails,
    double? weight,
    double? volume,
    bool? requiresSignature,
    bool? requiresPhoto,
    bool? requiresIdCheck,
    int? priority,
    DateTime? scheduledDeliveryTime,
    DateTime? deliveredAt,
    DateTime? eta,
    DeliveryStatus? status,
    String? signature,
    String? photo,
    String? notes,
  }) {
    return DeliveryPackage(
      id: id ?? this.id,
      trackingNumber: trackingNumber ?? this.trackingNumber,
      recipientName: recipientName ?? this.recipientName,
      recipientPhone: recipientPhone ?? this.recipientPhone,
      deliveryAddress: deliveryAddress ?? this.deliveryAddress,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      packageDetails: packageDetails ?? this.packageDetails,
      weight: weight ?? this.weight,
      volume: volume ?? this.volume,
      requiresSignature: requiresSignature ?? this.requiresSignature,
      requiresPhoto: requiresPhoto ?? this.requiresPhoto,
      requiresIdCheck: requiresIdCheck ?? this.requiresIdCheck,
      priority: priority ?? this.priority,
      scheduledDeliveryTime: scheduledDeliveryTime ?? this.scheduledDeliveryTime,
      deliveredAt: deliveredAt ?? this.deliveredAt,
      eta: eta ?? this.eta,
      status: status ?? this.status,
      signature: signature ?? this.signature,
      photo: photo ?? this.photo,
      notes: notes ?? this.notes,
    );
  }
}

class DeliveryEvent {
  final String id;
  final String packageId;
  final DeliveryStatus status;
  final DateTime timestamp;
  final Position location;
  final String? signature;
  final String? photo;
  final String? notes;
  
  DeliveryEvent({
    required this.id,
    required this.packageId,
    required this.status,
    required this.timestamp,
    required this.location,
    this.signature,
    this.photo,
    this.notes,
  });
}